src/dmroom.ts

```diff
@@ -0,0 +1,173 @@
+import { Client, DMChannel, GroupDMChannel, Message, User } from "discord.js";
+import { Intent} from "matrix-appservice-bridge";
+import { DbDmRoom } from "./db/dbdatadmroom";
+import { DMHandler } from "./dmhandler";
+import { Util } from "./util";
+import { Log } from "./log";
+
+const log = new Log("DMRoom");
+
+export class DMRoom {
+    private matrixMembers: Set<string>;
+    private sentMessages: Set<string>;
+    private deferLock: Promise<any>;
+    /* There are unfortunately a few differences between
+    the two chan types which means we will be doing a lot of
+    if elseing between them.*/
+    private channel: DMChannel|GroupDMChannel;
+
+    constructor(private dbroom: DbDmRoom, private handler: DMHandler) {
+        this.deferLock = Promise.resolve();
+        this.sentMessages = new Set();
+    }
+
+    public get DiscordChannelId() {
+        return this.dbroom.ChannelId;
+    }
+
+    public get RoomId() {
+        return this.dbroom.RoomId;
+    }
+
+    get discordUserIDs() {
+        if (this.channel.type === "dm") {
+            const c = (<DMChannel> this.channel);
+            return [c.client.user.id, c.recipient.id];
+        } else if (this.channel.type === "group") {
+            return (<GroupDMChannel> this.channel).recipients.keyArray();
+        }
+    };
+
+    public HydrateRoomWithDiscord(discordClient: Client) {
+        const channel = discordClient.channels.get(this.dbroom.ChannelId);
+        if (channel === undefined) {
+            log.warn("HydrateRoomWithDiscord was given a client that didn't have the channel.");
+            throw new Error("Channel not found for client!");
+        }
+        if (!["dm", "group"].includes(channel.type)) {
+            log.warn("HydrateRoomWithDiscord was given a channel that was not a dm|group.");
+            throw new Error("Channel is not a dm|group");
+        }
+        this.channel = <DMChannel|GroupDMChannel> channel;
+        log.verbose(`${this.dbroom.ChannelId} has ${this.discordUserIDs.length} discord members`);
+    }
+
+    public async HydrateRoomWithMatrix(matrixClient: Intent): Promise<void> {
+        const roomState: any[] = await matrixClient.roomState(this.dbroom.RoomId);
+        roomState.forEach((element) => {
+            if (element.type === "m.room.member" &&
+                element.content.membership === "join") {
```

Comment:
invite ghosts on failure etc...? Or do we leave that to the message sending fallback?

Reply:
Probs the fallback